File: spec/support/features/console.rb
class Console
  include Capybara::DSL

  def waiting?
    waiting_visible?(true)
  end

  # We need to verify using capybara asynchronous find, that there is
  # a waiting message on the console output and then that this waiting 
  # message is no longer visible .
  def has_waited?
    waiting?
    waiting_visible?(false)
  end

  def has_on_stdout?(output)
    has_output?('stdout', output)
  end

  def has_on_stderr?(output)
    has_output?('stderr', output)
  end

  def has_connection_error?
    find('#connect_error', visible: true)
  end

  def has_status?(status)
    has_output?('status', "[Program exited with status: #{status}]")
  end

  private

  def waiting_visible?(visible)
    find('#waiting', visible: visible)
  end

  def has_output?(stream, chunk)
    output.find(".#{stream}", text: chunk, match: :first)
  end

  def output
    find('.output')
  end
end

File: app/assets/javascripts/ground/console.js
function Console() {
    this.output = $('.output');
    this.connectError = $('#connect_error');
    this.waiting = $('#waiting');
}

Console.prototype.clean = function() {
    this.connectError.hide();
    this.output.find('span').each(function() {
        this.remove();
    });
};

Console.prototype.startWaiting = function() {
    this.clean();
    this.waiting.show();
};

Console.prototype.stopWaiting = function() {
    this.waiting.hide();
};

Console.prototype.write = function(stream, chunk) {
    switch (stream) {
        case 'status':
            this.stopWaiting();
            chunk = '[Program exited with status: ' + chunk + ']';
            break;
        case 'error':
            stream = 'stderr';
            this.stopWaiting();
            this.clean();
            break;
    }
    this.output.append($('<span class="'+stream+'">').text(chunk));
};

Console.prototype.error = function(error) {
    this.stopWaiting();
    this.clean();
    this.connectError.show();
};
